server/join.js

```diff
@@ -1,8 +1,9 @@
-const { curryN, tap } = require('ramda')
+const { curry } = require('ramda')
 
-// join : Socket -> (a -> String) -> a -> a
-const join = curryN(2, (socket, room) => tap(data =>
-  socket.join(room(data))
-))
+// join :: (Action -> String) -> Action -> Action
+const join = (room, axn) => {
+  axn.meta.socket.join(room(axn))
```

Comment:
In the past we've had direct access to the `socket` by registering `socket.io` middleware, like this:
```js
const someMiddleware = (socket, done) => {
  // build functions with socket
  done()
}
```
But that led to very stateful patterns.  So what I've proposed is to use `mount` to create a single `socket.io` middleware that lifts the entire `app`.
```js
const app = handle({ /* maps of types to handlers */ })
require('socket.io')(server).use(mount({ app }))
```
This means you no longer have that `socket`, because we don't create the middleware manually.  The solution I came up with was to stick it in the `action.meta` on every request.  Then if you carry it along (easy to do with `overPayload`), it will be available to `join` or `to` or whatever else you need.

Rooms are problematic, but at some level we need to be able to push data to specific users.  We can roll our own, but at the end of the day, they're still rooms.

We can discuss more offline if you like.  I'm open to suggestions.

Reply:
> The most common use case that I've seen is to broadcast to all sockets except the current one.

I think broadcasting to `io` is more common in some services.

---

I understand the motivation for this change. My concern is that the contract `to(room, axn)` is too ambiguous for my taste. It's not clear what socket we would broadcast to or that i need to preserve `meta.socket` for the duration of my transformation. 

I very much favor the old contract--possibly rearranging some args. That way, we could do the following explicitly:

```js
axn => to(axn.meta.socket, room, axn)
axn => to(io, room, axn)
axn => to(io, room, { type: 'STATIC_ACTION' }) // ignore `axn`
```

---

Actually, in all honesty, i think i'd recommend dropping explicit support for rooms entirely & discourage their use. They are so very stateful & don't fit well into this suspiciously redux-like pattern. Projects that are in the need of rooms should be able to integrate them adhoc with vanilla `socket.io`. I suspect this to be an unpopular viewpoint, however...